src/StreamingDataFrames/streamingdataframes/state/rocksdb/stores.py

```diff
@@ -0,0 +1,432 @@
+        self._batch = rocksdict.WriteBatch(raw_mode=True)
+        self._prefix = b""
+        self._failed = False
+        self._completed = False
+        self._dumps = dumps
+        self._loads = loads
+
+    @contextlib.contextmanager
+    def with_prefix(self, prefix: Any = b"") -> Self:
+        """
+        Prefix all the keys in the given scope.
+
+        Normally, it's called by Streaming DataFrames engine to ensure that every
+        message key is stored separately.
+
+        The `with_prefix` calls should not be nested.
+        Only one prefix can be set at a time.
+
+        :param prefix: a prefix string to be used.
+            Should be either `bytes` or JSON-serializable object.
+            The prefix doesn't need to contain the separator, it will be added
+            automatically between the key and the prefix itself if the prefix
+            is present.
+        """
+        if self._prefix != _DEFAULT_PREFIX:
+            raise NestedPrefixError("The transaction already has a prefix")
+        self._prefix = (
+            prefix if isinstance(prefix, bytes) else self._serialize_value(prefix)
+        )
+        if self._prefix:
+            self._prefix += _PREFIX_SEPARATOR
+
+        try:
+            yield self
+        finally:
+            self._prefix = b""
+
+    def get(self, key: Any, default: Any = None) -> Optional[Any]:
+        """
+        Get a key from the store.
+
+        It first looks up the key in the update cache in case it has been updated
+        but not flushed yet.
+
+        It returns `None` if the key is not found and `default` is not provided.
+
+        :param key: a JSON-deserializable key.
+        :param default: value to return if the key is not present in the state.
+            It can be of any type.
+        :return: a JSON-deserialized value
+        """
+        self._validate_transaction_state()
+
+        # First, check the update cache in case the value was previously written
+        # Use sentinel as default because the actual value can be "None"
+        key_serialized = self._serialize_key(key)
+        cached = self._update_cache.get(key_serialized, _sentinel)
+        if cached is not _sentinel:
+            return self._deserialize_value(cached)
+
+        # The value is not found in cache, check the db
+        stored = self._storage.get(key_serialized, _sentinel)
+        if stored is not _sentinel:
+            return self._deserialize_value(stored)
+        return default
+
+    def set(self, key: Any, value: Any):
+        """
+        Set a key to the store.
+
+        It first updates the key in the update cache.
+
+        :param key: a JSON-deserializable key.
+        :param value: a JSON-serializable value
+        """
+        self._validate_transaction_state()
+
+        key_serialized = self._serialize_key(key)
+        value_serialized = self._serialize_value(value)
+
+        try:
+            self._batch.put(key_serialized, value_serialized)
```

Comment:
The main reason I did it this way was the "delete" operation, yes.

As you noticed, we'd need to come up with a smarter way of keeping track of deleted keys, which seems to me an overcomplication in this case. 
Plus, we'd still spend +- the same memory because in the end you still do `batch.delete()`.

Reply:
Yeah it doesn't really make sense or help with any additional memory with deleting...I also don't know a better way of managing deletes, so I think everything you did makes sense, including matching `put` behavior with `delete`.

It still might be worth considering handling the `put`s slightly differently though since it's is probably the majority of usage. Just depends on if you think the slight difference in behavior is worth it. I also am not 100% sure it would make a difference, but it seems like it should.

Reply:
From the top of my head, the only case when it makes a difference is when you set the same keys many times during a single transaction.
Then you might have multiple "put" entries in WriteBatch for a single key in the update cache.
But it doesn't seem like a frequent situation, and I don't think it's worth optimizing for right now.